sound/soc/sof/control.c

```diff
@@ -110,6 +121,15 @@ int snd_sof_volume_put(struct snd_kcontrol *kcontrol,
 					    SOF_CTRL_CMD_VOLUME,
 					    true);
 
+	/*
+	 * Point scontrol back to cache location. Copy sent control data
+	 * if he IPC was successful.
+	 */
+	scontrol->control_data = pcdata;
+	if (!ret)
+		memcpy(pcdata->chanv, cdata->chanv, chanv_size);
```

Comment:
@singalsu This looks somewhat convoluted to me. I suppose we have bot .put() and .get(), for each .put(), we just send the target values to FW directly, and then we will use .get() to get the final values from FW directly and update it to application user, so cache here actually no needed?

Even the .put() setting fails in FW, the FW itself must using some other real values, using .get() to get the real values make sense?

Reply:
> @singalsu This looks somewhat convoluted to me. 

I didn't change IPC principle with this PR just fixed the scontrols cache corrupt bug with ext bytes put() and a hard to predict impact to FW configuration with get(). Those make use of sof-eqctl tool strange and likely also other user space originated control. Then the same fixed/improved IPC flow was extended to other controls. I tried to get feedback if this approach is OK or does it need a larger redesign. It's not cleanest possible I agree.

The kernel request/response buffers could be separate and the FW should not respond with random bytes the size of request if there's a fail. But that would be larger change.

>I suppose we have bot .put() and .get(), for each .put(), we just send the target values to FW directly, and then we will use .get() to get the final values from FW directly and update it to application user, so cache here actually no needed?

The scontrols cache is solely for PM restore to work. It uses for FW configuration restore the last put() or get() data. The PM suspend is not doing any context store for audio components. If there would be no PM suspending the scontrols cache would not be needed.

> 
> Even the .put() setting fails in FW, the FW itself must using some other real values, using .get() to get the real values make sense?

This code does not do get() after a failed put(). The assumption is that the last successful put() command can be use to restore firmware configuration.

Did this address your concern?

Reply:
@keyonjie ?